src/lib/raft/raft.c

```diff
@@ -480,15 +551,9 @@ raft_process_msg(struct raft *raft, const struct raft_msg *req, uint32_t source)
 		if (source == raft->leader) {
 			say_info("RAFT: the node %u has resigned from the "
 				 "leader role", raft->leader);
-			/*
-			 * Candidate node clears leader implicitly when starts a
-			 * new term, but non-candidate won't do that, so clear
-			 * it manually.
-			 */
-			raft->leader = 0;
-			raft_schedule_broadcast(raft);
-			if (raft->is_candidate)
-				raft_sm_schedule_new_election(raft);
+			raft_clear_leader(raft);
+			raft_ev_timer_stop(raft_loop(), &raft->timer);
```

Comment:
Ok, seems I've made it all work. Check out the diff below.
A couple of extra changes: looks like `raft_leader_unsee` is unneeded. I've inlined it everywhere already.
Also I made it so the leader doesn't see itself. This only complicated the resign logic: the leader sends us a resign request, and it also sends us the fact that it "doesn't see the leader". Let's make it so that only the followers signal each other whether they see the leader or not.

Notice, `raft_ev_timer_stop` is still needed for a non-candidate. Otherwise it'll continue tracking the resigned leader and waiting for its death.


<details><summary>diff</summary>

```diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index 9e724f334..a945608f8 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -412,26 +412,24 @@ raft_process_recovery(struct raft *raft, const struct raft_msg *req)
 static bool
 raft_leader_see(struct raft *raft);
 
-/** Perform needed actions on leader loss. */
-static bool
-raft_leader_unsee(struct raft *raft);
-
 /**
  * Clear leader-related fields on leader loss or resign and let everyone know
  * this node doesn't see the leader anymore.
  */
 static inline void
-raft_clear_leader(struct raft *raft)
+raft_leader_resign(struct raft *raft)
 {
-	if (raft->leader == raft->self) {
-		/*
-		 * Update leader_last_seen when resigning so that leader_idle
-		 * starts counting from zero after the resign.
-		 */
-		raft->leader_last_seen = raft_ev_monotonic_now(raft_loop());
-	}
+	assert(raft->leader == raft->self);
+	/*
+	 * Update leader_last_seen when resigning so that leader_idle
+	 * starts counting from zero after the resign.
+	 */
+	raft->leader_last_seen = raft_ev_monotonic_now(raft_loop());
 	raft->leader = 0;
-	raft_leader_unsee(raft);
+	bool was_seen = bit_test(&raft->leader_witness_map, raft->self);
+	assert(!was_seen);
+	(void)was_seen;
+	raft_schedule_broadcast(raft);
 }
 
 int
@@ -551,8 +549,21 @@ raft_process_msg(struct raft *raft, const struct raft_msg *req, uint32_t source)
 		if (source == raft->leader) {
 			say_info("RAFT: the node %u has resigned from the "
 				 "leader role", raft->leader);
-			raft_clear_leader(raft);
+			/*
+			 * Candidate node clears leader and stops the timer
+			 * implicitly when starts a new term, but non-candidate
+			 * won't do that, so do it all manually.
+			 */
+			raft->leader = 0;
+			bit_clear(&raft->leader_witness_map, raft->self);
 			raft_ev_timer_stop(raft_loop(), &raft->timer);
+			raft_schedule_broadcast(raft);
+			/*
+			 * No need for pre-vote checks when the leader
+			 * deliberately told us it's resigning.
+			 */
+			if (raft->is_candidate)
+				raft_sm_schedule_new_election(raft);
 		}
 		return 0;
 	}
@@ -744,9 +755,6 @@ raft_sm_become_leader(struct raft *raft)
 	assert(!raft->is_write_in_progress);
 	raft->state = RAFT_STATE_LEADER;
 	raft->leader = raft->self;
-	bool changed = raft_leader_see(raft);
-	assert(changed);
-	(void)changed;
 	raft_ev_timer_stop(raft_loop(), &raft->timer);
 	/* State is visible and it is changed - broadcast. */
 	raft_schedule_broadcast(raft);
@@ -951,7 +959,7 @@ raft_sm_stop(struct raft *raft)
 	raft->is_enabled = false;
 	raft->is_candidate = false;
 	if (raft->state == RAFT_STATE_LEADER)
-		raft_clear_leader(raft);
+		raft_leader_resign(raft);
 	raft->state = RAFT_STATE_FOLLOWER;
 	raft_ev_timer_stop(raft_loop(), &raft->timer);
 	/* State is visible and changed - broadcast. */
@@ -1046,7 +1054,7 @@ raft_stop_candidate(struct raft *raft)
 	raft->is_candidate = false;
 	if (raft->state == RAFT_STATE_LEADER) {
 		assert(!raft_ev_timer_is_active(&raft->timer));
-		raft_clear_leader(raft);
+		raft_leader_resign(raft);
 	} else if (!raft_is_leader_seen(raft)) {
 		raft_ev_timer_stop(raft_loop(), &raft->timer);
 	} else {
@@ -1107,20 +1115,6 @@ raft_leader_see(struct raft *raft)
 	return false;
 }
 
-static bool
-raft_leader_unsee(struct raft *raft)
-{
-	uint32_t source = raft->self;
-	assert(source > 0 && source < VCLOCK_MAX);
-	bool was_seen = bit_clear(&raft->leader_witness_map, source);
-	if (was_seen) {
-		raft_sm_election_update(raft);
-		raft_schedule_broadcast(raft);
-		return true;
-	}
-	return false;
-}
-
 void
 raft_cfg_is_candidate_later(struct raft *raft, bool is_candidate)
 {


```

</details>

Reply:
Another amendment, since the leader doesn't "see" itself anymore.
```diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index a945608f8..a5732ba7e 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -1095,7 +1095,11 @@ raft_notify_is_leader_seen(struct raft *raft, bool is_leader_seen,
                was_seen = bit_set(&raft->leader_witness_map, source);
        } else {
                was_seen = bit_clear(&raft->leader_witness_map, source);
-               if (was_seen)
+               /*
+                * Ignore notifications on whether someone sees this instance,
+                * if it's the leader.
+                */
+               if (was_seen && raft->state != RAFT_STATE_LEADER)
                        raft_sm_election_update(raft);
        }
        return was_seen != is_leader_seen;
```